File: api/catalog.py
# -*- coding: utf-8 -*-
"""GET /api/catalog?tag=hanime&page=1 -> { status, tag, page, totalPages, count, titles[] }"""
import re

from _lib import parse_request, json_resp, err, cors_headers, http_get, BASE


def handler(request):
    if parse_request(request)[2] == "OPTIONS":
        return {"statusCode": 204, "headers": cors_headers(), "body": ""}

    _, query, _ = parse_request(request)
    tag = (query.get("tag") or "hanime").strip().lower()
    try:
        page = max(1, int(query.get("page") or "1"))
    except ValueError:
        page = 1

    page_url = f"{BASE}/tag/{tag}/" if page == 1 else f"{BASE}/tag/{tag}/page/{page}/"
    html = http_get(page_url)
    if not html:
        return err("Failed to fetch tag page", 502)

    titles = []
    seen = set()
    card_re = re.compile(
        r'id="manga-item-(\d+)"[\s\S]*?href="https://hentaihaven\.xxx/watch/([^"/]+)/"[^>]*title="([^"]+)"[\s\S]*?<img[^>]*src="([^"]+)"'
    )
    for m in card_re.finditer(html):
        pid, slug, title, poster = m.group(1), m.group(2), m.group(3), m.group(4)
        if slug in seen:
            continue
        seen.add(slug)
        block = html[m.start() : m.start() + 6000]
        eps = set(re.findall(r"watch/[^\"/]+/episode-(\d+)", block))
        titles.append(
            {
                "id": int(pid),
                "title": title.strip(),
                "slug": slug,
                "poster": poster,
                "url": f"{BASE}/watch/{slug}/",
                "episodes": len(eps),
                "episodeNumbers": sorted(int(e) for e in eps),
            }
        )

    if not titles:
        return err("No entries found (tag may not exist)", 404)

    lp = re.search(r'page/(\d+)/"[^>]*>\s*»', html)
    total_pages = int(lp.group(1)) if lp else page
    return json_resp(
        {"status": True, "tag": tag, "page": page, "totalPages": total_pages, "count": len(titles), "titles": titles}
    )

File: api/catalog.ts
import { corsHeaders, json, error, fetchHtml } from './_lib';

/**
 * GET /api/catalog?tag=hanime&page=1
 *
 * Parses a hentaihaven.xxx tag page into a title list.
 * The default tag "hanime" mirrors hanime.tv content.
 * Other tags work too (creampie, uncensored, ahegao, ...).
 */
export default async function handler(req: Request) {
  if (req.method === 'OPTIONS') return new Response(null, { status: 204, headers: corsHeaders() });

  const url = new URL(req.url);
  const tag = (url.searchParams.get('tag') || 'hanime').trim().toLowerCase();
  const page = Math.max(1, parseInt(url.searchParams.get('page') || '1', 10) || 1);

  const pageUrl =
    page === 1
      ? `https://hentaihaven.xxx/tag/${encodeURIComponent(tag)}/`
      : `https://hentaihaven.xxx/tag/${encodeURIComponent(tag)}/page/${page}/`;

  const html = await fetchHtml(pageUrl);
  if (!html) return error('Failed to fetch tag page', 502);

  const titles: any[] = [];
  const seen = new Set<string>();

  // Each series card: <div id="manga-item-<id>" ...><a href=".../watch/<slug>/" title="Title"><img src="poster"></a> ... episode links
  const cardRe = /id="manga-item-(\d+)"[\s\S]*?href="https:\/\/hentaihaven\.xxx\/watch\/([^"/]+)\/"[^>]*title="([^"]+)"[\s\S]*?<img[^>]*src="([^"]+)"/g;
  let m: RegExpExecArray | null;
  while ((m = cardRe.exec(html))) {
    const [, id, slug, title, poster] = m;
    if (seen.has(slug)) continue;
    seen.add(slug);

    // count episodes in the same card block (episode-N links)
    const cardStart = m.index;
    const cardEnd = html.indexOf('</div>', cardStart) + 6;
    const cardBlock = html.slice(cardStart, cardStart + 6000);
    const eps = new Set<string>();
    const epRe = /watch\/[^"/]+\/episode-(\d+)/g;
    let em: RegExpExecArray | null;
    while ((em = epRe.exec(cardBlock))) eps.add(em[1]);

    titles.push({
      id: Number(id),
      title: title.trim(),
      slug,
      poster,
      url: `https://hentaihaven.xxx/watch/${slug}/`,
      episodes: eps.size,
      episodeNumbers: Array.from(eps).map(Number).sort((a, b) => a - b),
    });
  }

  if (titles.length === 0) return json({ status: false, error: 'No entries found (tag may not exist)' }, 404);

  // pagination info
  const lastPage = html.match(/page\/(\d+)\/"\s*class="[^"]*last[^"]*"/) || html.match(/page\/(\d+)\/"[^>]*>\s*»/);
  const totalPages = lastPage ? parseInt(lastPage[1], 10) : page;

  return json({ status: true, tag, page, totalPages, count: titles.length, titles });
}
